### ISSUE TITLE
Gaps

### ISSUE THREAD

**Issue by LevavasseurYann** [chuanconggao/PrefixSpan-py#5]
Hi,

That's not a real issue i'm just wondering if exist any way to set a maximum and/or minimum gaps for the pattern mining ?

**Comment by pintergreg** [comment]
@chuanconggao What about adding the starting position of the sequence to the matches tuple? 
For example:
[1, 2, 2] [(2, **1**, 3), (3, **2**, 4)]
[1, 3, 4] [(0, **1**, 4), (1, **2**, 4)]
[1, 1, 1] [(1, **0**, 2), (3, **0**, 2)]

Then the difference of the third and the second number could be the gap.

**Comment by chuanconggao** [comment]
@pintergreg Thanks for the feedback. I will use a simple sequence as an example of why a new algorithm is required to support gap.

Imagine a sequence `a a a b`. Assume the pattern `a b` is frequent, the existing algorithms under the Pattern-Growth framework from PrefixSpan find `a b`, where `a` matches the first `a` and `b` matches the first `b`. However, if we have a constraint of maximum gap of 0, then `a b` can only be found when `a` matches the third `a`. 

Like you suggested, we can record the matching positions. In fast, we are already recording but not reporting. However, it can only tell you the gap of the mined patterns. It cannot guarantee finding all the patterns when applying  any constraint of minimum/maximum gap.

The major issue is that the current algorithm framework does not consider the second and third `a` as the start of pattern, if there is already the first `a` before them as the start of pattern. Thus, people propose new algorithms like Gap-BIDE. Arguable, the new algorithms are still based on Pattern-Growth framework, and I can update the framework accordingly. However, as the current framework supports not only one algorithm, but three algorithms, it has to be done very carefully. To make things worse, the logics of checking and pruning of closed/generator patterns have to be carefully updated too.

In summary, we can support gap, maybe starting from PrefixSpan, then to BIDE and FEAT. For BIDE, there is already an algorithm Gap-BIDE which updated the logics of checking and pruning. However, for FEAT, there is no existing algorithm to refer to. As I designed FEAT, probably I will design the new Gap-FEAT algorithm.  

I do not have a clear timeline right now. It should take me a full day to add support to PrefixSpan.

**Comment by pintergreg** [comment]
@chuanconggao Thank you for the detailed explanation. Now, I understand better how it works.
Actually, I was afraid that the problem is to distinguish all of the `a`-s ad select the one matches the gap constrain.

**Comment by raisauku** [comment]
@chuanconggao I was wondering if you have added the gap support to PrefixSpan or is it still considered a future work? Thanks

**Comment by chuanconggao** [comment]
Hi @raisauku, I do plan to add this nice feature. But this does require extra effort and will likely take at least one week from me. Right now, there is no specific date.

My top priority is fixing the issue of BIDE algorithm for correctness. This feature is the next one in my future plan.

**Comment by wenpiip** [comment]
Hi @chuanconggao , have you added gap constraints to prefixspan now? According to the paper https://www.researchgate.net/.../221506141_Generalization_of_Pattern-Growth_Method... , instead of looking only for the first occurrence of element, every element's occurrence is considered when construct the projected database. So are you gonna use their method for the gap constraint or you have your own method? Plus, does your prefix span work for item-set based element (for example for sequence (ac)(bcd)(ab), where each element have several items).
